### ISSUE TITLE
Question: Keep track of invalid agent test cases?

### ISSUE THREAD

**Issue by jaydiablo** [browscap/browscap#1291]
I had asked about adding a "failure" test (a test that makes sure a useragent is parsed as "Default Browser") for Samsung Browser versions 3.1, 4.1, 5.1 since it seems like those versions don't exist (https://github.com/browscap/browscap/pull/1286#issuecomment-287621173).

I was curious how much this is done in browscap currently, and it seems like there have been a few tickets in the past that have test cases that do this:

>tests/fixtures/issues/issue-084.php
tests/fixtures/issues/issue-267.php
tests/fixtures/issues/issue-295.php
tests/fixtures/issues/issue-436.php
tests/fixtures/issues/issue-497.php
tests/fixtures/issues/issue-498.php
tests/fixtures/issues/issue-500.php
tests/fixtures/issues/issue-513.php
tests/fixtures/issues/issue-523.php
tests/fixtures/issues/issue-539.php
tests/fixtures/issues/issue-620.php
tests/fixtures/issues/issue-687.php
tests/fixtures/issues/issue-725.php
tests/fixtures/issues/issue-749.php
tests/fixtures/issues/issue-784.php
tests/fixtures/issues/issue-811.php
tests/fixtures/issues/issue-814.php
tests/fixtures/issues/issue-821.php
tests/fixtures/issues/issue-849.php
tests/fixtures/issues/issue-865.php
tests/fixtures/issues/issue-900.php
tests/fixtures/issues/issue-904.php

Would it be helpful to maintain a single fixture that stores all of these sorts of "invalid ua" test cases, perhaps with a comment (where possible) describing why that particular pattern should be considered invalid?

**Comment by mimmi20** [comment]
In our sources we can't add comments, because the json format doesnt support them. But we could add a test file with a suffix in the name.

**Comment by jaydiablo** [comment]
@mimmi20 I was thinking of something like a test fixture (PHP file) that's named something like `issue-000-invalids.php`. In that file, above each test case, the reason the agent is invalid could be a comment, i.e:

```
    // Samsung Browser 5.1 was never released: http://url/for/reason
    'issue-000-invalids-A' => [
```

**Comment by mimmi20** [comment]
That sounds good.

**Comment by jaydiablo** [comment]
As mentioned in the comments of #1331, it seems like the "fake" patterns were removed at some point.  I assume the "fakeSafari", "fakeIE" browsers don't exist anymore with that change.

That said, what should be browscap's stance on the agents that are fake?  Currently there are some in the test suite, should these remain at all (I remove one in #1331 since it was causing some issues with the changes I was making)?

Is there any value in detecting that these are fake (it seems like it's tricky to maintain patterns for this, which I assume is why the fake patterns were removed). Seems like it would be nice to prevent fake patterns from ending up in the test suite, unless they're specifically in a file like the "invalids" file I added in #1298, but I suppose that's maybe left to the maintainers to not add obviously fake patterns to the suite rather than specifically testing that fake patterns return a certain result or something like that.

Sorry if this has been discussed before, just thought it was relevant to this ticket and something that came up when I was doing the work for #1331.

**Comment by mimmi20** [comment]
I removed these pattern because they created too much false matches.

**Comment by jaydiablo** [comment]
So I have a question about some patterns that are testing that they be "Default Browser".

As part of #1344 I have a list of Useragents like this:

>Opera/8.50 (Macintosh; U; fr)
Opera/8.51 (Macintosh; U; fr)
Opera/8.52 (Macintosh; U; da)
Opera/8.53 (Macintosh; U; la)
Opera/8.54 (Macintosh; U; br)
Opera/7.0 (Macintosh; U; es)
Opera/7.10 (Macintosh; U; br)
Opera/7.11 (Macintosh; U; ko)
Opera/7.51 (Macintosh; U; ru)
Opera/9.12 (Macintosh; U; br)
Opera/9.80 (Macintosh; U; cs)
Opera/6.01 (Windows; U; tu)
Opera/6.04 (Windows XP; U; [de])
Opera/7.21 (Windows; U; nb)
Opera/7.22 (Windows; U; nl)
Opera/7.52 (Windows; U; fi)
Opera/8.02 (Windows; U; lt)
Opera/8.50 (Windows; U; pt)
Opera/8.51 (Windows; U; cs)
Opera/8.52 (Windows; U; ca)
Opera/8.53 (Windows; U; br)
Opera/8.54 (Windows; U; es)
Opera/9.12 (Windows; U; hu)
Opera/9.80 (Windows; U; lt)

It's easy enough to add some patterns that will catch these properly, which I did, but then realized that it was causing some tests in the suite to fail.

These agents typically parse as "Opera" in other parsers (http://useragent.mkf.solutions/?userAgent=Opera%2F8.51+%28Windows%3B+U%3B+cs%29) but that certainly doesn't mean that the agents aren't fake, or should be parsed in the same way.

Browscap has several test cases setup to detect these formats as "Default Browser":

 * issue-900-AUE
 * issue-900-AUF
 * issue-900-AUG
 * issue-900-AUI
 * issue-900-AUJ
 * issue-900-AUY
 * issue-900-AVC

(I can't link to these directly since that fixture file is so big)

Were these added to the test suite to detect as "Default Browser" because they're known fake strings that we don't want Browscap to pick up?  If so, I'll move them to the invalid issue fixture with a suitable comment.

If not, I can update the tests to what they'll parse as with these changes I have prepared.

FWIW, there are some patterns that match similar strings, just different platforms:

`Opera?*(#PLATFORM#)*` 
https://github.com/browscap/browscap/blob/master/resources/user-agents/browsers/opera/opera-generic.json#L207
https://github.com/browscap/browscap/blob/master/resources/user-agents/browsers/opera/opera-generic.json#L199
https://github.com/browscap/browscap/blob/master/resources/user-agents/browsers/opera/opera-generic.json#L183
https://github.com/browscap/browscap/blob/master/resources/user-agents/browsers/opera/opera-8-00.json#L166

etc...

**Comment by mimmi20** [comment]
Maybe these useragents (which are parsed as "Default Browser") were part of the removed fake files. If they were marked as fake, then it was I did not think the simply "Windows" or "Macintosh" were Not-Fake-Platform pattern.
I did not see any useragent like these in the logs of the server I have access to (I did only see later versions in the Mozilla form).

If other useragent parsers parse these Uas as Opera we should that too.
